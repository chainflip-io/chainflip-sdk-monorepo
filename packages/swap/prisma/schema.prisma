// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

// `pnpm prisma format` after changes :)

generator client {
  provider        = "prisma-client-js"
  // https://github.com/prisma/prisma/issues/2584#issuecomment-646414549
  // https://github.com/prisma/prisma/issues/6603#issuecomment-849979849
  output          = "../node_modules/.prisma/client"
  previewFeatures = ["multiSchema"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
  schemas  = ["public", "private"]
}

enum Chain {
  Polkadot
  Ethereum
  Bitcoin

  @@schema("public")
}

enum Asset {
  FLIP
  USDC
  DOT
  ETH
  BTC

  @@schema("public")
}

model SwapDepositChannel {
  id                    BigInt   @id @default(autoincrement())
  uuid                  String   @unique @default(uuid())
  srcAsset              Asset
  depositAddress        String
  expectedDepositAmount Decimal  @db.Decimal(30, 0)
  destAsset             Asset
  destAddress           String
  expiryBlock           Int
  issuedBlock           Int
  createdAt             DateTime @default(now())
  swaps                 Swap[]

  @@index([depositAddress])
  @@schema("public")
}

model Swap {
  id                        BigInt              @id @default(autoincrement())
  /// the swap id on the state chain
  nativeId                  BigInt              @unique
  srcAsset                  Asset
  depositAmount             Decimal              @db.Decimal(30, 0)
  intermediateAmount        Decimal?             @db.Decimal(30, 0)
  egressAmount              Decimal?             @db.Decimal(30, 0)
  destAsset                 Asset
  destAddress               String
  depositReceivedAt         DateTime
  depositReceivedBlockIndex String
  swapExecutedAt            DateTime?
  swapExecutedBlockIndex    String?
  egressCompletedAt         DateTime?
  egressCompletedBlockIndex String?
  swapDepositChannel        SwapDepositChannel? @relation(fields: [swapDepositChannelId], references: [id])
  swapDepositChannelId      BigInt?
  txHash                    String?             @unique
  egress                    Egress?
  createdAt                 DateTime            @default(now())
  updatedAt                 DateTime            @default(now()) @updatedAt

  @@schema("public")
}

model Egress {
  id          BigInt     @id @default(autoincrement())
  /// the egress id on the state chain
  nativeId    BigInt
  /// the swap id on the state chain
  chain       Chain
  amount      Decimal    @db.Decimal(30, 0)
  timestamp   DateTime
  swap        Swap?      @relation(fields: [swapId], references: [id])
  swapId      BigInt?    @unique
  broadcast   Broadcast? @relation(fields: [broadcastId], references: [id])
  broadcastId BigInt?
  createdAt   DateTime   @default(now())

  @@unique([nativeId, chain])
  @@schema("public")
}

model Broadcast {
  id       BigInt   @id @default(autoincrement())
  /// the broadcast id on the state chain
  nativeId BigInt
  chain    Chain
  egresses Egress[]

  @@unique([nativeId, chain])
  @@schema("public")
}

model State {
  id     Int @id @default(autoincrement())
  height Int @default(0)

  @@schema("private")
}

model MarketMaker {
  id        Int    @id @default(autoincrement())
  name      String @unique
  publicKey String

  @@schema("private")
}

enum ThirdPartyProtocol {
  lifi
  squid

  @@schema("public")
}

model ThirdPartySwap {
  id            BigInt             @id @default(autoincrement())
  uuid          String             @unique
  protocol      ThirdPartyProtocol
  routeResponse Json
  txHash        String
  txLink        String
  createdAt     DateTime           @default(now())

  @@schema("public")
}
