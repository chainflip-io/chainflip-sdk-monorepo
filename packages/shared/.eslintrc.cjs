// eslint-disable-next-line @typescript-eslint/no-var-requires
const path = require('path');

module.exports = {
  extends: '../../.eslintrc.json',
  plugins: ['eslint-plugin-n'],
  rules: {
    'n/no-process-env': ['error'],
    'import/no-extraneous-dependencies': [
      'error',
      {
        packageDir: [__dirname, path.join(__dirname, '..', '..')],
      },
    ],
  },
  overrides: [
    {
      files: ['*.test.ts', '*.mjs', '*.js', '*.cjs'],
      rules: {
        'n/no-process-env': ['off'],
      },
    },
  ],
};
