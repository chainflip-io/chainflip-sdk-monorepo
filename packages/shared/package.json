{
  "name": "@chainflip-io/sdk-shared",
  "version": "1.0.0",
  "description": "",
  "scripts": {
    "typecheck": "tsc --noEmit",
    "clean": "rm -rf dist",
    "build": "tsc -p tsconfig.cjs.json && tsc -p tsconfig.mjs.json",
    "build:abis": "typechain --target=ethers-v5 --out-dir=./src/abis ./abis/**",
    "lint": "pnpm eslint --max-warnings 0 './**/*.ts'",
    "test": "jest",
    "test:ci": "CI=1 pnpm test -- --forceExit"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "peerDependencies": {
    "zod": "^3.22.2"
  },
  "dependencies": {
    "@polkadot/util": "^12.4.2",
    "@polkadot/util-crypto": "^12.4.2",
    "ethers": "5.7.2",
    "rxjs": "^7.8.1",
    "ws": "^8.13.0"
  },
  "devDependencies": {
    "@ethersproject/abi": "^5.7.0",
    "@ethersproject/providers": "^5.7.2",
    "@typechain/ethers-v5": "^11.1.1",
    "@types/ws": "^8.5.5",
    "typechain": "^8.3.1",
    "winston": "^3.10.0",
    "zod": "^3.22.2"
  }
}
