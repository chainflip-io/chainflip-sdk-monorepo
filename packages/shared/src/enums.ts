type ArrayToMap<T extends readonly string[]> = {
  [K in T[number]]: K;
};

const arrayToMap = <const T extends readonly string[]>(
  array: T,
): ArrayToMap<T> =>
  Object.fromEntries(array.map((key) => [key, key])) as ArrayToMap<T>;

export const InternalAssets = arrayToMap(['Flip', 'Usdc', 'Dot', 'Eth', 'Btc']);
export type InternalAsset =
  (typeof InternalAssets)[keyof typeof InternalAssets];

export const Chains = arrayToMap(['Bitcoin', 'Ethereum', 'Polkadot']);
export type Chain = (typeof Chains)[keyof typeof Chains];

export const Assets = arrayToMap(['FLIP', 'USDC', 'DOT', 'ETH', 'BTC']);
export type Asset = (typeof Assets)[keyof typeof Assets];

export const ChainflipNetworks = arrayToMap([
  'backspin',
  'sisyphos',
  'perseverance',
  'mainnet',
]);

export type ChainflipNetwork =
  (typeof ChainflipNetworks)[keyof typeof ChainflipNetworks];

export const isTestnet = (network: ChainflipNetwork): boolean =>
  network !== ChainflipNetworks.mainnet;

// https://github.com/chainflip-io/chainflip-backend/blob/a2a3c2e447e7b629c4b96797d9eed22eb5b87a0b/state-chain/primitives/src/chains/assets.rs#L51-L59
export const assetConstants = {
  [InternalAssets.Eth]: {
    chain: Chains.Ethereum,
    asset: Assets.ETH,
    name: 'Ether',
    decimals: 18,
    contractId: 1,
  },
  [InternalAssets.Flip]: {
    chain: Chains.Ethereum,
    asset: Assets.FLIP,
    name: 'FLIP',
    decimals: 18,
    contractId: 2,
  },
  [InternalAssets.Usdc]: {
    chain: Chains.Ethereum,
    asset: Assets.USDC,
    name: 'USDC',
    decimals: 6,
    contractId: 3,
  },
  [InternalAssets.Dot]: {
    chain: Chains.Polkadot,
    asset: Assets.DOT,
    name: 'Polkadot',
    decimals: 10,
    contractId: 4,
  },
  [InternalAssets.Btc]: {
    chain: Chains.Bitcoin,
    asset: Assets.BTC,
    name: 'Bitcoin',
    decimals: 8,
    contractId: 5,
  },
} satisfies Record<
  InternalAsset,
  {
    chain: Chain;
    asset: Asset;
    name: string;
    decimals: number;
    contractId: number;
  }
>;

// https://github.com/chainflip-io/chainflip-backend/blob/a2a3c2e447e7b629c4b96797d9eed22eb5b87a0b/state-chain/primitives/src/chains.rs#L52-L56
export const chainConstants = {
  [Chains.Ethereum]: {
    assets: [Assets.ETH, Assets.FLIP, Assets.USDC],
    gasAsset: Assets.ETH,
    contractId: 1,
  },
  [Chains.Polkadot]: {
    assets: [Assets.DOT],
    gasAsset: Assets.DOT,
    contractId: 2,
  },
  [Chains.Bitcoin]: {
    assets: [Assets.BTC],
    gasAsset: Assets.BTC,
    contractId: 3,
  },
} satisfies Record<
  Chain,
  { assets: Asset[]; gasAsset: Asset; contractId: number }
>;

export type AssetOfChain<C extends Chain> =
  (typeof chainConstants)[C]['assets'][number];

export type UncheckedAssetAndChain = { asset: Asset; chain: Chain };
export type AssetAndChain = {
  [C in Chain]: { chain: C; asset: AssetOfChain<C> };
}[Chain];

export type ChainAssetMap<T> = {
  [C in Chain]: {
    [A in AssetOfChain<C>]: T;
  };
};

export function isValidAssetAndChain(
  assetAndChain: UncheckedAssetAndChain,
): assetAndChain is AssetAndChain {
  const { asset, chain } = assetAndChain;
  const validAssets = chainConstants[chain].assets as string[];

  return validAssets.includes(asset);
}

export function assertIsValidAssetAndChain(
  assetAndChain: UncheckedAssetAndChain,
): asserts assetAndChain is AssetAndChain {
  if (!isValidAssetAndChain(assetAndChain)) {
    throw new Error('invalid asset and chain combination');
  }
}

export const readChainAssetMap = <T>(
  value: ChainAssetMap<T>,
  asset: UncheckedAssetAndChain,
): T => {
  assertIsValidAssetAndChain(asset);
  const chainValues = value[asset.chain];
  return chainValues[asset.asset as keyof typeof chainValues];
};

export function getInternalAsset(asset: UncheckedAssetAndChain) {
  const map: ChainAssetMap<InternalAsset> = {
    [Chains.Ethereum]: {
      [Assets.USDC]: InternalAssets.Usdc,
      [Assets.FLIP]: InternalAssets.Flip,
      [Assets.ETH]: InternalAssets.Eth,
    },
    [Chains.Bitcoin]: {
      [Assets.BTC]: InternalAssets.Btc,
    },
    [Chains.Polkadot]: {
      [Assets.DOT]: InternalAssets.Dot,
    },
  };

  return readChainAssetMap(map, asset);
}
