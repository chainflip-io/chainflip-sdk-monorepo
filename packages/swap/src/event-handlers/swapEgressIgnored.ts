import { z } from 'zod';
import { u64, chainflipAssetEnum, u128, hexString } from '@/shared/parsers';
import { getStateChainError } from './common';
import { EventHandlerArgs } from './index';

const swapEgressIgnoredArgs = z.object({
  asset: chainflipAssetEnum,
  amount: u128,
  swapId: u64,
  reason: z.object({
    __kind: z.literal('Module'),
    value: z.object({
      error: hexString,
      index: z.number(),
    }),
  }),
});

export type SwapDepositAddressReadyEvent = z.input<
  typeof swapEgressIgnoredArgs
>;

export default async function swapEgressIgnored({
  prisma,
  event,
  block,
}: EventHandlerArgs): Promise<void> {
  const { amount, swapId, reason } = swapEgressIgnoredArgs.parse(event.args);

  const [failure, swap] = await Promise.all([
    getStateChainError(prisma, block, reason.value),
    prisma.swap.findUniqueOrThrow({ where: { nativeId: swapId } }),
  ]);

  await prisma.ignoredEgress.create({
    data: {
      swapId: swap.id,
      ignoredAt: new Date(block.timestamp),
      ignoredBlockIndex: `${block.height}-${event.indexInBlock}`,
      amount: amount.toString(),
      stateChainErrorId: failure.id,
    },
  });
}
