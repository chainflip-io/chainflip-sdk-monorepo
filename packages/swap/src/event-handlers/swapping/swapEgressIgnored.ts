import { swappingSwapEgressIgnored as schema160 } from '@chainflip/processor/160/swapping/swapEgressIgnored';
import { swappingSwapEgressIgnored as schema190 } from '@chainflip/processor/190/swapping/swapEgressIgnored';
import { z } from 'zod';
import { getStateChainError } from '../common.js';
import { EventHandlerArgs } from '../index.js';

const swapEgressIgnoredArgs = z.union([schema190, schema160]);

export type SwapEgressIgnoredArgs = z.input<typeof swapEgressIgnoredArgs>;

export default async function swapEgressIgnored({
  prisma,
  event,
  block,
}: EventHandlerArgs): Promise<void> {
  const { amount, swapRequestId, reason } = swapEgressIgnoredArgs.parse(event.args);

  const failure = await (reason.__kind === 'Module'
    ? getStateChainError(prisma, block, reason.value)
    : prisma.stateChainError.upsert({
        where: {
          specVersion_palletIndex_errorIndex: { specVersion: 0, palletIndex: 0, errorIndex: 0 },
        },
        create: {
          specVersion: 0,
          palletIndex: 0,
          errorIndex: 0,
          docs: 'Unknown error',
          name: 'Unknown error',
        },
        update: {},
      }));

  await prisma.ignoredEgress.create({
    data: {
      ignoredAt: new Date(block.timestamp),
      ignoredBlockIndex: `${block.height}-${event.indexInBlock}`,
      amount: amount.toString(),
      stateChainError: { connect: { id: failure.id } },
      swapRequest: { connect: { nativeId: swapRequestId } },
      type: 'SWAP',
    },
  });
}
