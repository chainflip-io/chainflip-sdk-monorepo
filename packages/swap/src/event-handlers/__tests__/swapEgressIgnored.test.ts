import { Assets } from '@/shared/enums';
import metadataMock from './metadata.json';
import {
  DOT_ADDRESS,
  createDepositChannel,
  swapEgressIgnoredMock,
} from './utils';
import prisma from '../../client';
import swapEgressIgnored from '../swapEgressIgnored';

jest.mock('axios', () => ({
  post: jest.fn().mockImplementation(async () => ({
    data: metadataMock,
  })),
}));

const {
  eventContext: { event },
  block,
} = swapEgressIgnoredMock;

describe(swapEgressIgnored, () => {
  beforeEach(async () => {
    await prisma.$queryRaw`TRUNCATE TABLE "SwapDepositChannel", "Swap", "Egress", "IgnoredEgress", "StateChainError" CASCADE`;
  });

  it('adds an IgnoredEgress record for a egress ignored event', async () => {
    const { swapId } = event.args;

    // store a new swap intent to initiate a new swap
    const swapDepositChannel = await createDepositChannel({
      swaps: {
        create: {
          nativeId: BigInt(swapId),
          depositAmount: '10000000000',
          swapInputAmount: '10000000000',
          swapOutputAmount: '10000000000',
          depositReceivedAt: new Date(block.timestamp - 12000),
          depositReceivedBlockIndex: `${block.height - 100}-${
            event.indexInBlock
          }`,
          swapExecutedAt: new Date(block.timestamp - 6000),
          swapExecutedBlockIndex: `${block.height}-${event.indexInBlock}`,
          srcAsset: Assets.ETH,
          destAsset: Assets.DOT,
          destAddress: DOT_ADDRESS,
          type: 'SWAP',
        },
      },
    });

    await prisma.$transaction((tx) =>
      swapEgressIgnored({
        block: block as any,
        event: event as any,
        prisma: tx,
      }),
    );

    const swap = await prisma.swap.findFirstOrThrow({
      where: { swapDepositChannelId: swapDepositChannel.id },
      include: { ignoredEgress: true },
    });

    expect(swap).toMatchSnapshot({
      id: expect.any(BigInt),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
      swapDepositChannelId: expect.any(BigInt),
      ignoredEgress: {
        id: expect.any(BigInt),
        swapId: expect.any(BigInt),
        stateChainErrorId: expect.any(Number),
        createdAt: expect.any(Date),
      },
    });

    expect(
      await prisma.stateChainError.findUniqueOrThrow({
        where: { id: swap.ignoredEgress!.stateChainErrorId },
      }),
    ).toMatchSnapshot({
      id: expect.any(Number),
    });
  });
});
