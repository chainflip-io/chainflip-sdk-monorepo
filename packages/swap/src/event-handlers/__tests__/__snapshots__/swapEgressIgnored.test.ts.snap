// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`swapEgressIgnored adds an IgnoredEgress record for a egress ignored event 1`] = `
{
  "ccmDepositReceivedBlockIndex": null,
  "ccmGasBudget": null,
  "ccmMessage": null,
  "createdAt": Any<Date>,
  "depositAmount": "10000000000",
  "depositReceivedAt": 2022-12-06T14:31:33.000Z,
  "depositReceivedBlockIndex": "20-0",
  "destAddress": "1yMmfLti1k3huRQM2c47WugwonQMqTvQ2GUFxnU7Pcs7xPo",
  "destAsset": "DOT",
  "egressId": null,
  "id": Any<BigInt>,
  "ignoredEgress": {
    "amount": "11",
    "createdAt": Any<Date>,
    "id": Any<BigInt>,
    "ignoredAt": 2022-12-06T14:31:45.000Z,
    "ignoredBlockIndex": "120-0",
    "stateChainErrorId": Any<Number>,
    "swapId": Any<BigInt>,
  },
  "intermediateAmount": null,
  "nativeId": 2n,
  "srcAsset": "ETH",
  "swapDepositChannelId": Any<BigInt>,
  "swapExecutedAt": 2022-12-06T14:31:39.000Z,
  "swapExecutedBlockIndex": "120-0",
  "swapInputAmount": "10000000000",
  "swapOutputAmount": "10000000000",
  "txHash": null,
  "type": "SWAP",
  "updatedAt": Any<Date>,
}
`;

exports[`swapEgressIgnored adds an IgnoredEgress record for a egress ignored event 2`] = `
{
  "docs": "The amount is below the minimum egress amount.",
  "errorIndex": 6,
  "id": Any<Number>,
  "name": "bitcoinIngressEgress.BelowEgressDustLimit",
  "palletIndex": 34,
  "specVersion": 1,
}
`;
